Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/distance.h

```diff
@@ -62,121 +76,719 @@ triangle_grid_sampling(const typename Kernel::Point_3& p0,
+      }
+      out = CGAL::cpp11::copy_n(g, nb_points, out);
+    }
+  }
+
   return out;
 }
 
-template <class Kernel, class TriangleRange1, class TriangleRange2>
-double approximated_Hausdorff_distance(
-  const TriangleRange1& triangles_1,
-  const TriangleRange2& triangles_2,
-  double targeted_precision
-)
+template<class OutputIterator, class TriangleMesh>
+OutputIterator
+sample_triangle_mesh(const TriangleMesh& tm,
+                           OutputIterator out)
+{
+  return sample_triangle_mesh(tm, out, parameters::all_default());
+}
+
+template <class Concurrency_tag,
+          class Kernel,
+          class PointRange,
+          class TriangleMesh,
+          class VertexPointMap>
+double approximate_Hausdorff_distance(
+  const PointRange& original_sample_points,
+  const TriangleMesh& tm,
+  VertexPointMap vpm)
 {
-  std::vector<typename Kernel::Point_3> sample_points;
-  sample_triangles<Kernel>( triangles_1,
-                            targeted_precision,
-                            std::back_inserter(sample_points) );
-  // std::ofstream out("/tmp/samples.xyz");
-  // std::copy(sample_points.begin(), sample_points.end(), std::ostream_iterator<typename Kernel::Point_3>(out,"\n"));
+  CGAL_assertion_code(  bool is_triangle = is_triangle_mesh(tm) );
+  CGAL_assertion_msg (is_triangle,
+        "Mesh is not triangulated. Distance computing impossible.");
+  #ifdef CGAL_HAUSDORFF_DEBUG
+  std::cout << "Nb sample points " << sample_points.size() << "\n";
+  #endif
+  typedef typename Kernel::Point_3 Point_3;
+  std::vector<Point_3> sample_points
+    (boost::begin(original_sample_points), boost::end(original_sample_points) );
 
-  typedef typename TriangleRange2::const_iterator Triangle_iterator;
-  typedef AABB_triangle_primitive<Kernel, Triangle_iterator> Primitive;
-  typedef AABB_traits<Kernel, Primitive> Traits;
-  typedef AABB_tree< Traits > Tree;
+  spatial_sort(sample_points.begin(), sample_points.end());
 
-  Tree tree(triangles_2.begin(), triangles_2.end());
-  tree.accelerate_distance_queries();
+  typedef AABB_face_graph_triangle_primitive<TriangleMesh> Primitive;
+  typedef AABB_tree< AABB_traits<Kernel, Primitive> > Tree;
 
-  double hdist = 0;
-  BOOST_FOREACH(const typename Kernel::Point_3& pt, sample_points)
+  Tree tree( faces(tm).first, faces(tm).second, tm);
+  tree.accelerate_distance_queries();
+  tree.build();
+  Point_3 hint = get(vpm, *vertices(tm).first);
+#ifndef CGAL_LINKED_WITH_TBB
+  CGAL_static_assertion_msg (!(boost::is_convertible<Concurrency_tag, Parallel_tag>::value),
+                             "Parallel_tag is enabled but TBB is unavailable.");
+#else
+  if (boost::is_convertible<Concurrency_tag,Parallel_tag>::value)
+  {
+    cpp11::atomic<double> distance(0);
+    Distance_computation<Tree, Point_3> f(tree, hint, sample_points, &distance);
+    tbb::parallel_for(tbb::blocked_range<std::size_t>(0, sample_points.size()), f);
+    return distance;
```

Comment:
@maxGimeno @sloriot This code is not the best one to compute a maximum using thread.

Currently:
  - the code uses `tbb::parallel_for`,
  - all threads compute a "local" maximum,
  - then they compare to a shared `atomic<double>` to synchronize the value.

Note that:
  - `atomic<double>` is not guaranteed to be lock-free,
  - there is a compile error with clang+C++20, in recent test-results:
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-161/Polygon_mesh_processing/TestReport_lrineau_ArchLinux-clang-CXX20-Release.gz

I think the code should use [`tbb::parallel_reduce`](https://software.intel.com/en-us/node/506154): each thread returns a value, and the main thread is responsible for the reduction of all values to the final value.